src/multiversioning.zig

```diff
@@ -0,0 +1,892 @@
+            releases: []const u32,
+            checksums: []const u128,
+            offsets: []const u32,
+            sizes: []const u32,
+            visits: []const bool,
+        }) PastReleases {
+            assert(past_init.releases.len == count);
+            assert(past_init.checksums.len == count);
+            assert(past_init.offsets.len == count);
+            assert(past_init.sizes.len == count);
+            assert(past_init.visits.len == count);
+
+            assert(count <= past_releases_max);
+            assert(count > 0);
+
+            for (past_init.releases) |v| assert(v != 0);
+            assert(std.sort.isSorted(u32, past_init.releases, {}, std.sort.asc(u32)));
+
+            assert(past_init.offsets[0] == 0);
+            for (past_init.offsets[1..], 0..) |offset, i| {
+                const calculated_offset = blk: {
+                    var calculated_offset: u32 = 0;
+                    for (past_init.sizes[0..i]) |size| {
+                        calculated_offset += size;
+                    }
+                    break :blk calculated_offset;
+                };
+                assert(offset != 0);
+                assert(offset == calculated_offset);
+            }
+
+            for (past_init.sizes) |s| assert(s != 0);
+
+            var past_releases = PastReleases{};
+            past_releases.count = count;
+
+            stdx.copy_disjoint(.inexact, u32, &past_releases.releases, past_init.releases);
+            stdx.copy_disjoint(.inexact, u128, &past_releases.checksums, past_init.checksums);
+            stdx.copy_disjoint(.inexact, u32, &past_releases.offsets, past_init.offsets);
+            stdx.copy_disjoint(.inexact, u32, &past_releases.sizes, past_init.sizes);
+            stdx.copy_disjoint(.inexact, bool, &past_releases.visits, past_init.visits);
+
+            return past_releases;
+        }
+    };
+
+    /// Covers MultiversionHeader[@sizeOf(u128)..].
+    checksum_header: u128 = undefined,
+
+    /// The AEGIS128L checksum of the binary, if the header (`.tb_mvh`) section were zeroed out.
+    /// Used to validate that the binary itself is not corrupt. Putting this in requires a bit
+    /// of trickery:
+    /// * inject a zero `.tb_mvh` section of the correct size,
+    /// * compute the hash,
+    /// * update the section with the correct data.
+    /// Used to ensure we don't try and exec into a corrupt binary.
+    checksum_binary_without_header: u128 = 0,
+
+    /// The current release is executed differently to past releases embedded in the body, so store
+    /// it separately. See exec_latest vs exec_release.
+    current_release: u32,
+
+    /// The AEGIS128L checksum of the direct output of `zig build`, for the current_release, before
+    /// any objcopy or build magic has been performed.
+    /// Used when extracting binaries from past releases to ensure a hash chain - this will be put
+    /// in the entry in past.checksums for this release by the build script.
```

Comment:
> Used when extracting binaries from past releases to ensure a hash chain

Sorry to revisit this, but I realized I still don't understand what this means. :thinking: 

By "hash chain" I expected something like
```
A = hash(a)
B = hash(b, A)
C = hash(c, B)
...
```
but I'm not spotting something like that -- each binary is hashed independently (afact). We keep a list of those hashes in `PastReleases.checksums`, but they aren't cumulative.

What am I missing?

Reply:
Ahh that makes sense, thank you!